### PR TITLE
VB-2729 Establishment specific contact information Telephone address

### PR DESCRIPTION
**Pull request by AledWEvans**
This is to add **telephone address** to the contact details

Background

Currently, establishment specific contact information is held in the front end of the service. This is because we had originally sent text messages directly from the front end of the service to inform visitors of bookings, updates and cancellations.

With the introduction of the admin service we can now move this information to the back end service - this will enable us to have better control of the information and allow us to onboard establishments quicker as we will not require information to be added to front end and released each time.

### CODE REVIEW

File: src/test/kotlin/uk/gov/justice/digital/hmpps/prisonregister/integration/emailaddress/DeletePrisonEmailResourceTest.kt [hunk 1]
```diff
@@ -73,16 +72,15 @@ class DeletePrisonEmailResourceTest : IntegrationTest() {
     // Given
     val prisonId = "BRI"
     val departmentType = SOCIAL_VISIT
-    val endPoint = "/secure/prisons/id/$prisonId/department/${departmentType.pathVariable}/email-address"
-
+    val endPoint = getEndPoint(prisonId, departmentType)
     // When
```
[the remark this answers is not in the document]
Comment (AledWEvans) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/prisonregister/integration/emailaddress/DeletePrisonEmailResourceTest.kt:98]: tests added

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/resource/validator/ValidTelephoneAddress.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+class TelephoneAddressValidator : ConstraintValidator<ValidTelephoneAddress, String> {
+  override fun isValid(value: String?, context: ConstraintValidatorContext?): Boolean {
+    if (value.isNullOrBlank()) return false
+    return this.isValidTelephoneAddress(value)
+  }
+  private fun isValidTelephoneAddress(value: String): Boolean {
+    return try {
+      // If no + prefix defaults to uk
+      val countryCode = if (value.startsWith("+")) "" else "GB"
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/resource/validator/ValidTelephoneAddress.kt:28]: is a non UK phone number valid?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/service/PrisonService.kt [hunk 3]
```diff
@@ -127,36 +134,90 @@ class PrisonService(
...
+  fun setTelephoneAddress(prisonId: String, newTelephoneAddress: String, departmentType: DepartmentType): SetOutcome {
     val contactDetails = contactDetailsRepository.getByPrisonIdAndType(prisonId, departmentType)
+    if (contactDetails == null) {
+      val prison = prisonRepository.getReferenceById(prisonId) ?: throw EntityNotFoundException()
+      val persistedTelephoneAddress = createOrGetTelephoneAddress(newTelephoneAddress)
+      contactDetailsRepository.saveAndFlush(ContactDetails(prisonId, prison, departmentType, telephoneAddress = persistedTelephoneAddress))
+      return SetOutcome.CREATED
+    }
+    val oldTelephoneAddress = contactDetails.emailAddress?.value
+    if (oldTelephoneAddress != newTelephoneAddress) {
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/service/PrisonService.kt:165]: shall we move update when existing to a different method?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/resource/validator/ValidTelephoneAddress.kt (hunk 2 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/resource/validator/ValidTelephoneAddress.kt:28]: Well was not going to get into this, as long as it's a tel number,.. better than no check at all,.. I was thinking do we cover prisons out side the normal +44,...

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/service/PrisonService.kt [hunk 4]
```diff
@@ -127,36 +134,108 @@ class PrisonService(
...
+    }
+  }
+
   @Transactional
-  @PreAuthorize(CLIENT_CAN_MAINTAIN_EMAIL_ADDRESSES)
+  @PreAuthorize(CLIENT_CAN_MAINTAIN_ADDRESSES)
   fun deleteEmailAddress(prisonId: String, departmentType: DepartmentType, throwNotFound: Boolean = false) {
     val contactDetails = contactDetailsRepository.getByPrisonIdAndType(prisonId, departmentType)
-
     contactDetails?.let {
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/service/PrisonService.kt:201]: minor comment - 201 and 202 can be combined, same below.

Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonregister/service/PrisonService.kt:201]: I will leave it I think it's more readable really..

**Review by danbenton-mojdt — APPROVED** [review]

**Pull request merged**